src/calendar_duration.rs

```diff
@@ -0,0 +1,294 @@
+        Some(self)
+    }
+
+    /// Sets the accurate component of this duration to a value in seconds.
+    ///
+    /// This duration will count an exact number of seconds, which includes the counting of leap
+    /// seconds.
+    ///
+    /// The minutes component will be set to zero.
+    ///
+    /// # Errors
+    ///
+    /// Returns `None` if the value of the seconds component would be out of range.
+    #[must_use]
+    pub const fn with_seconds(mut self, seconds: u64) -> Option<Self> {
+        self.mins_and_secs = try_opt!(MinutesAndSeconds::from_seconds(seconds));
+        Some(self)
+    }
+
+    /// Sets the nanoseconds component of this duration to `nanos`.
+    ///
+    /// # Errors
+    ///
+    /// Returns `None` if `nanos` exceeds 999_999_999.
+    pub const fn with_nanos(mut self, nanos: u32) -> Option<Self> {
+        if nanos >= 1_000_000_000 {
+            return None;
+        }
+        self.nanos = nanos;
+        Some(self)
+    }
+
+    /// Sets the nanoseconds component of this duration to `micros * 1000`.
+    ///
+    /// # Errors
+    ///
+    /// Returns `None` if `micros` exceeds 999_999.
+    pub const fn with_micros(self, micros: u32) -> Option<Self> {
+        if micros >= 1_000_000 {
+            return None;
+        }
+        self.with_nanos(micros * 1000)
+    }
+
+    /// Sets the nanoseconds component of this duration to `millis * 1000`.
+    ///
+    /// # Errors
+    ///
+    /// Returns `None` if `millis` exceeds 999.
+    pub const fn with_millis(self, millis: u32) -> Option<Self> {
+        if millis >= 1000 {
+            return None;
+        }
+        self.with_nanos(millis * 1_000_000)
+    }
+
+    /// Returns the `months` component of this calendar duration.
+    pub const fn months(&self) -> u32 {
+        self.months
+    }
+
+    /// Returns the `days` component of this calendar duration.
+    pub const fn days(&self) -> u32 {
+        self.days
+    }
+
+    /// Returns the `minutes` and `seconds` components of this calendar duration.
+    pub const fn mins_and_secs(&self) -> (u64, u64) {
+        self.mins_and_secs.mins_and_secs()
+    }
+
+    /// Returns the `nanos` component of this calendar duration.
+    pub const fn nanos(&self) -> u32 {
+        self.nanos
+    }
+
+    /// Returns `true` if this `CalendarDuration` spans no time.
+    pub const fn is_zero(&self) -> bool {
+        let (mins, secs) = self.mins_and_secs();
+        self.months == 0 && self.days == 0 && mins == 0 && secs == 0 && self.nanos == 0
+    }
+}
+
+/// Type to encode either seconds, or minutes and up to 60 seconds.
+///
+/// We don't allow having both an `u64` with minutes and an `u64` with seconds.
+/// Having two components in a `CalendarDuration` whose only subtle difference is how they behave
+/// around leap seconds will make the type unintuitive.
+///
+/// # Encoding
+///
+/// - Seconds:                                     `seconds << 2 | 0b10`
```

Comment:
How much memory is this stuff saving? It's not obvious to me that this is worth it.

Reply:
My main motivation is to encode the either *seconds* or *minutes and seconds* cases.

Still I prefer to make the type not larger than necessary, especially if one of the goals is to replace the `Days` and `Months` types. An enum will add 8 bytes to a type of currently 20 bytes; 40%.

Reply:
I spent some time thinking this over and would still prefer that we start with a "real" enum -- that doesn't rely on bit-shifting tricks. If we get an issue that complains about the size we can decide whether it's worth optimizing this. But it seems unlikely to me at this point that people will want to hold a ton of these in an `Vec` or similar.

Reply:
With an enum we would still have most of the complexity.
- In my opinion we want to keep the range checks in `MinutesAndSeconds::from_seconds` and `from_minutes_and_seconds`.
  The restricted range for minutes gives as the guarantee `minutes * 60 + seconds` does not overflow.
  The checks would be needed to allow for this 'optimization' in the future.
- We still need the special case where we encode a value created with `from_minutes_and_seconds(0, s)` the same as `from_seconds(s)` so that they compare equal.
- We still need a match to get both values in `mins_and_secs()`.

Switching to an enum is not going to save many lines or maybe even none.

I just don't consider a couple of bitwise operations much of a problem and well worth it for the much nicer type size. Really don't want to drop them :disappointed:.

Reply:
How are we going to proceed here?

Reply:
I still think this is a textbook case of premature optimization and think we should drop the bitshifting for now.

Reply:
We have a difference of opinion here. If chrono gets a `CalendarDuration` as new functionality I want it to be as good as I can make it. Do you want to block the functionality on this detail?